presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -166,11 +168,36 @@ public synchronized void addLocation(URI location, TaskId remoteSourceTaskId)
                 scheduler,
                 pageBufferClientCallbackExecutor);
         allClients.put(location, client);
+        checkState(taskIdToURI.put(remoteSourceTaskId, location) == null, "Duplicate remoteSourceTaskId: " + remoteSourceTaskId);
```

Comment:
Potentially there can be a race, when remote location is removed before it is added. We should return from `addLocation` when the client is already cancelled.